tmc-langs-framework/src/main/java/fi/helsinki/cs/tmc/langs/domain/Filer.java

```diff
@@ -40,100 +39,80 @@ public Filer setLanguagePlugin(LanguagePlugin languagePlugin) {
-//        System.out.println("relativePath: " + relativePath);
-//        System.out.println("");
-        logger.info("Looking into file: {} ", file);
-        try {
-            if (skipFile(file)) {
-                logger.info("Skipping file: {} ", file);
-                return;
-            }
-            Path toFile = toPath.resolve(relativePath);
-            if (justCopy(file)) {
-                Files.createDirectories(toFile.getParent());
-                Files.copy(file, toFile, StandardCopyOption.REPLACE_EXISTING);
-                logger.info("Just copying file from: {} to:{}", file, toFile);
-            } else {
-                List<String> data = readFile(file);
-                List<String> output = prepareFile(data);
-                if (!output.isEmpty()) {
-                    Files.createDirectories(toFile.getParent());
-                    Files.write(toFile, output, StandardCharsets.UTF_8, StandardOpenOption.CREATE);
-                    logger.info("Filtered file while copying from: {} to:{}", file, toFile);
-                } else {
-                    logger.info("skipped file as empty while copying from: {} to:{}", file, toFile);
-                }
+    protected final boolean isTextFile(Path file) throws IOException {
+        List<String> skipList = Arrays.asList(new String[] {"class", "jar"});
+        if (file.toFile().isFile() && skipList.contains(getFileExtension(file))) {
+            return false;
+        }
+        // If >90% of characters in file are sourcecode-text-like, interpret as a text file
+        long countNormal = 0;
+        long countOther = 0;
+        List<String> data = readFile(file);
+        for (String line : data) {
+            for (int i=0; i<line.length(); i++) {
+                char c = line.charAt(i);
+                if (c >= 32 && c <= 125) countNormal++; // 0-9,A-Z,a-z,{-]+!...
+                else if (c == 9)         countNormal++; // horizontal tab
+                else                     countOther++;
             }
-        } catch (IOException ex) {
-            throw new IllegalStateException(ex);
         }
+        if (countOther == 0) return true;
+        return (countNormal / countOther > 9);
     }
-
+    
+    
     /*
      * Note: we need to use the inputStream for reading more interesting filetypes.
      */
-    private List<String> readFile(Path file) {
+    protected List<String> readFile(Path file) throws IOException {
         List<String> data = new ArrayList<>();
-        try (Scanner scanner = new Scanner(Files.newInputStream(file))) {
-            while (scanner.hasNextLine()) {
-                data.add(scanner.nextLine());
-            }
-        } catch (IOException ex) {
-            throw new IllegalStateException(ex);
+        Scanner scanner = new Scanner(Files.newInputStream(file));
+        while (scanner.hasNextLine()) {
+            data.add(scanner.nextLine());
```

Comment:
Since we are in any case reading the whole file into memory, just use [`java.nio.file.Files.readAllLines(Path path, Charset cs)`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#readAllLines(java.nio.file.Path,%20java.nio.charset.Charset)).

If memory becomes an issue this should be changed to a stream, but my intuition says this is unlikely considering the average file sizes and the relatively high amounts of memory in modern systems.